src/test/java/org/sonatype/nexus/repository/composer/internal/ComposerContentFacetImplTest.java

```diff
@@ -0,0 +1,282 @@
+
+  private static final String PACKAGES_PATH = "packages.json";
+
+  private static final String PROVIDER_PATH = "p/vendor/project.json";
+
+  private static final String ZIPBALL_PATH = "vendor/project/version/project-version.zip";
+
+  @Mock
+  private Component component;
+
+  @Mock
+  private Asset asset;
+
+  @Mock
+  private BlobRef blobRef;
+
+  @Mock
+  private Blob blob;
+
+  @Mock
+  private InputStream blobInputStream;
+
+  @Mock
+  private Repository repository;
+
+  @Mock
+  private StorageFacet storageFacet;
+
+  @Mock
+  private Bucket bucket;
+
+  @Mock
+  private StorageTx tx;
+
+  @Mock
+  private Content upload;
+
+  @Mock
+  private NestedAttributesMap assetAttributes;
+
+  @Mock
+  private NestedAttributesMap contentAttributes;
+
+  @Mock
+  private NestedAttributesMap cacheAttributes;
+
+  @Mock
+  private NestedAttributesMap formatAttributes;
+
+  @Mock
+  private AssetBlob assetBlob;
+
+  @Mock
+  private TempBlob tempBlob;
+
+  private ComposerContentFacetImpl underTest;
+
+  @Before
+  public void setUp() throws Exception {
+    underTest = new ComposerContentFacetImpl(COMPOSER_FORMAT);
+    underTest.attach(repository);
+
+    when(tx.findBucket(repository)).thenReturn(bucket);
+    when(tx.requireBlob(blobRef)).thenReturn(blob);
+    when(tx.createAsset(bucket, COMPOSER_FORMAT)).thenReturn(asset);
+    when(tx.createAsset(bucket, component)).thenReturn(asset);
+    when(tx.findComponents(any(Query.class), eq(singletonList(repository)))).thenReturn(emptyList());
+    when(tx.createComponent(bucket, COMPOSER_FORMAT)).thenReturn(component);
+
+    when(repository.facet(StorageFacet.class)).thenReturn(storageFacet);
+
+    when(asset.attributes()).thenReturn(assetAttributes);
+    when(asset.formatAttributes()).thenReturn(formatAttributes);
+    when(asset.requireBlobRef()).thenReturn(blobRef);
+    when(asset.requireContentType()).thenReturn(CONTENT_TYPE);
+    when(asset.getChecksums(HASH_ALGORITHMS)).thenReturn(CHECKSUMS);
+
+    when(assetAttributes.child("cache")).thenReturn(cacheAttributes);
+    when(assetAttributes.child("composer")).thenReturn(formatAttributes);
+    when(assetAttributes.child("content")).thenReturn(contentAttributes);
+
+    when(assetBlob.getBlob()).thenReturn(blob);
+    when(assetBlob.getBlobRef()).thenReturn(blobRef);
+
+    when(contentAttributes.get("last_modified", Date.class)).thenReturn(LAST_MODIFIED);
+    when(contentAttributes.get("etag")).thenReturn(ETAG);
+
+    when(cacheAttributes.get("last_verified", Date.class)).thenReturn(LAST_VERIFIED);
+
+    when(storageFacet.createTempBlob(upload, HASH_ALGORITHMS)).thenReturn(tempBlob);
+
+    when(blob.getInputStream()).thenReturn(blobInputStream);
+
+    when(upload.getContentType()).thenReturn(CONTENT_TYPE);
+
+    when(component.group(any(String.class))).thenReturn(component);
+    when(component.name(any(String.class))).thenReturn(component);
+    when(component.version(any(String.class))).thenReturn(component);
```

Comment:
Minor: It looks like these setups are grouped into logical groups, it would make it clearer to future developers if they were just extracted into methods i.e. setupComponentBehaviour.

Reply:
> Minor: It looks like these setups are grouped into logical groups, it would make it clearer to future developers if they were just extracted into methods i.e. setupComponentBehaviour.

I don't know if I buy that; to me I'd rather see all the setup in one place if possible, so long as it's no bigger than one screen or so (and if it's bigger than that, probably an indication that we need to refactor the underlying code).

For now since I need this for the next PR I'm just going to merge this one down, but I have no objection to doing a cleanup PR on this later. You're a better guardian of the style on Formats at this point than me since you're over there at the moment.

Reply:
Fine here 👍